Add a title guess method to get "better" title

leaf-soba opened the pull request:
- Closes:#11999
- As the issue said I should "traditional" (non-AI) code, so I write a AI-liked code in a simple traditional way. Since I don't have enough data, so the rules is limited now.
- I didn't add it `PdfMergeMetadataImporterTest` now, since I'm not sure if we like this kind of method or not, so I want to set up the PR with a demo code to discuss it first. I'll add it to `PdfMergeMetadataImporter` with a better clear code if we like it.
<!-- 
Describe the changes you have made here: what, why, ... 
Link the issue that will be closed, e.g., "Closes #333".
If your PR closes a koppor issue, link it using its URL, e.g., "Closes https://github.com/koppor/jabref/issues/47".
"Closes" is a keyword GitHub uses to link PRs with issues; do not change it.
Don't reference an issue in the PR title because GitHub does not support auto-linking there.
-->

### Mandatory checks

<!-- 
- Go through the list below. Please don't remove any items.
- [x] done; [ ] not done / not applicable
-->

- [x] I own the copyright of the code submitted and I licence it under the [MIT license](https://github.com/JabRef/jabref/blob/main/LICENSE)
- [ ] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [x] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [ ] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.

koppor reviewed (changes requested):
Good to have a test case.

But I think, there should be some earlier switch. See the hint `stripper.setSortByPosition(true);`

Please try again longer how to handle two-column PDFs. -- The title should be parsed corectly by `org.jabref.logic.importer.fileformat.PdfContentImporter#getEntryFromPDFContent`

What I meant is: Do two passes: One with sorted by coordinates and one keeping text blocks together. The abstract is probably better prased with keeping text blocks together; the title better with position.

One can tweak org.jabref.logic.importer.fileformat.PdfContentImporter#getEntryFromPDFContent to see if the paper is a single-column or two column paper. One could even go further and detect if it is Springer, IEEE or another format.

leaf-soba commented:
> Good to have a test case.
> 
> But I think, there should be some earlier switch. See the hint `stripper.setSortByPosition(true);`
> 
> Please try again longer how to handle two-column PDFs. -- The title should be parsed corectly by `org.jabref.logic.importer.fileformat.PdfContentImporter#getEntryFromPDFContent`
> 
> What I meant is: Do two passes: One with sorted by coordinates and one keeping text blocks together. The abstract is probably better prased with keeping text blocks together; the title better with position.
> 
> One can tweak org.jabref.logic.importer.fileformat.PdfContentImporter#getEntryFromPDFContent to see if the paper is a single-column or two column paper. One could even go further and detect if it is Springer, IEEE or another format.

Wow, I thought I wrote something cool and AI-like method, but it looks like it’s not quite what we need! 😅 I’ve now added a method to grab the title by position. This is just a demo code to see if the approach works for us—there’s definitely room for improvement, especially in formatting and commented code. If we’re happy with it, I’ll clean things up and refactor the org.jabref.logic.importer.fileformat.PdfContentImporter#getEntryFromPDFContent input!

leaf-soba commented:
> > * Since I don't have enough data, so the rules is limited now.
> 
> Please list the files available for others to be able to review.
> 
> I think, JabRef should have an example file for the first four templates listed at https://latextemplates.github.io/.
> 
> If not, pleae add.
> 
> I see following dirctories containing pre-formatted files
> 
> * https://github.com/JabRef/jabref/tree/main/src/test/resources/org/jabref/logic/importer/util
> * https://github.com/JabRef/jabref/tree/main/src/test/resources/pdfs/IEEE
> 
> And some more at https://github.com/JabRef/jabref/tree/main/src/test/resources/pdfs - where I think, most of them do not follow a publisher format.
> 
> I thought that first dowing a parsing regarding position sorted for title and author, then reparsing with onsorted, but with blocks for abstract is the most working approach.
> 
> First, the format of the PDF should be guessed. If IEEE, then set parameters for IEEE. If LNCS, set format for LNCS.
> 
> Guessing the title by a rectangle works if you know the publisher format. Thus, that guessing should go first.
> 
> For 563 TB of test data - outside Europe and US and maybe many other countries, go to https://annas-archive.org/. -- You can also start with 32 TB of test data.
> 
> ![image](https://private-user-images.githubusercontent.com/1366654/378429725-d09b36ff-1192-4a25-a932-5d9aa0f89a1f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk1MTU0ODUsIm5iZiI6MTcyOTUxNTE4NSwicGF0aCI6Ii8xMzY2NjU0LzM3ODQyOTcyNS1kMDliMzZmZi0xMTkyLTRhMjUtYTkzMi01ZDlhYTBmODlhMWYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAyMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMjFUMTI1MzA1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MTQwYmNlNDdlOGI4NDJmMGQwMmMyYzM4YmQ1ZGJhNjFlM2RlMjY5MTJiYTIwOTExNjFiZWI3ZTBmYjU4ZjhmNCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.6mXCtinWdHV4Zs2JzeU9s7pw6d9zTlbvrVgl1lUZoIs)

Sorry, so far I have only tested it with the single file mentioned in the issue, se2paper.pdf. I haven’t written academic papers in years, so I’m not very familiar with the IEEE and LNCS formats. I’ll do my best to learn and research them from the test data tomorrow

leaf-soba commented:
I added 4 more unit test case with "IEEE", "LNCS", "scientificThesis", and a "thesis-example" I found in the pdf folder, new code can pass them all so far, if we need more test I can add it.
The new idea is the title should be the largest font size in most of papers which I checked in the test data.

koppor commented:
Remove commented code. You have it in your git history if you ever need it.

leaf-soba commented:
> Remove commented code. You have it in your git history if you ever need it.

sorry it is a dirty code now, I want to set up a demo code to check if we like this solution or not, if the solution confirmed I'll refactor it.

koppor commented:
Please propose a PR to the user documentation at https://docs.jabref.org/collect/findunlinkedfiles#pdfs-for-which-it-works - explaining the larger titles.

koppor commented:
I am currently very busy, thus quick replies only. Sorry for that.

> I added 4 more unit test case with "IEEE", "LNCS", "scientificThesis", and a "thesis-example" I found in the pdf folder, 

Sounds good.

The idea using the largest object is smart! For future work, maybe, the second page needs to be checked, too. Sometimes, there is a cover page in front. I currently do not have an example at hand, but I don't have an example at hand - and I think, the heuristics will get harder then.

> if we need more test I can add it. 

Can you add follwing papers as test cases:

1. https://link.springer.com/article/10.1007/s10664-023-10367-y
2. https://link.springer.com/article/10.1007/s10664-020-09875-y --> note that it would be great if the sub title was parsed
3. https://onlinelibrary.wiley.com/doi/10.1002/spe.3169
4. https://peerj.com/articles/cs-213/#
5. https://dl.acm.org/doi/10.1145/3597503.3639130 (ACM format)

When adding, add a `README.md` to the folder where there are stored with the source of the PDF. I mean, a table with filename as first column and link as second column.

---

In all these papers, you can also test for the abstract.

In all thesse papers, you can test for the authors (which is probably a brain-teaser and left for future work).

koppor commented:
Follow-up (copied from the docs): The next development step is to extract the title of the PDF, use the "Lookup DOI" functionality and then merge that information. -- I think, there is already much code existing using the merge, but the lookup DOI is not done yet. Could be a three-liner.

leaf-soba commented:
I add 5 more unittest and the old code can pass 9 of them, 1 of them(s10664-023-10367-y,"Do RESTful API design rules have an impact on the understandability of Web APIs?") have a special icon in the bottom right with an extra large font size character in it. I check some test data most of titles should not be in the place 10% bottom, so I change my code a little to fix it.
ToDo in this PR:
1. add a README.md to the folder where there are stored with the source of the PDF.
2. refactor my code to make the logic clear.
3. extract the title of the PDF, use the "Lookup DOI" functionality and then merge that information.(Or it should be in a new PR?)
ToDo in a new PR:
1. get abstract in a new PR
2. get authors in a new PR
3. get subTitle in a new PR
4. propose a PR to the user documentation https://docs.jabref.org/collect/findunlinkedfiles#pdfs-for-which-it-works explaining the larger titles.

koppor commented on src/main/java/org/jabref/logic/importer/fileformat/PdfContentImporter.java:
```diff
@@ -208,8 +210,83 @@ public ParserResult importDatabase(Path filePath) {
+
     // make this method package visible so we can test it
-    Optional<BibEntry> getEntryFromPDFContent(String firstpageContents, String lineSeparator) {
+    Optional<BibEntry> getEntryFromPDFContent(String firstpageContents, String lineSeparator, String titleByPosition) {
```
After thinking, allowing `null` or `""` is OK. Everthing else is harder.

koppor commented on src/main/java/org/jabref/logic/importer/fileformat/PdfContentImporter.java:
```diff
@@ -437,7 +514,8 @@ Optional<BibEntry> getEntryFromPDFContent(String firstpageContents, String lineS
         if (title != null) {
-            entry.setField(StandardField.TITLE, title);
+//            title = guessBetterTitleInMetaData(buildMetaData(author, editor, abstractT, keywords, title, conference, doi, series, volume, number, pages, year, publisher));
+            entry.setField(StandardField.TITLE, titleByPosition == null || titleByPosition.isEmpty() ? title : titleByPosition);
```
No, this is flawed

- If there is no title - the titleByPosition should be used.
- Use `org.jabref.model.strings.StringUtil#isNullOrEmpty` for clear checking
- Split up `?`in real `if/else`
- Do a check for `!StringUtil.isNullOrEmpty` first, then store this in `title` field.
- Then, the if `title != null` and setting the field can be kept. ( ` entry.setField(StandardField.TITLE, title`)

leaf-soba commented:
> If you see something. Otherwise, just remove commented-out code (and adress my comments)
> We need to get this in soon, because we have > 50 pull requests open - and all reviewers loose context. If this is around for weeks, we start from scratch at revewieing.
Definitely, I’ll start by removing the commented-out code and addressing your feedback tomorrow.
Apologies for the delay—since I’ve been balancing this with my full-time job, I worked on it in smaller parts. I understand this has taken more time, and I appreciate your patience. While I’m still getting familiar with this project, I’m comfortable with code reviews, so please feel free to assign any tasks to help move things forward.

leaf-soba commented on src/main/java/org/jabref/logic/importer/fileformat/PdfContentImporter.java:
```diff
@@ -485,7 +563,7 @@ private String getFirstPageContents(PDDocument document) throws IOException {
         stripper.setStartPage(1);
         stripper.setEndPage(1);
-        stripper.setSortByPosition(true);
+//        stripper.setSortByPosition(true);
```
I apologize for forgetting to uncomment this code. Whether or not there is a stripper.setSortByPosition(true); does not impact my unit tests.

leaf-soba commented:
1. revert the temp change of unit test to original one.
2. all of title's character should stay together, add a `isFarAway` method to it to pass the unit test for hello world case.
3. change guess title variable name form old version `titleByPosition` to `titleByFontSize`
4. remove all commented code.
5. Add a `@VisibleForTesting` to `getEntryFromPDFContent`.
6. rewrite the javaDoc for `getEntryFromPDFContent`
7. fix the logic issue for setting title.

leaf-soba commented on src/test/java/org/jabref/logic/importer/fileformat/PdfContentImporterTest.java:
```diff
@@ -31,7 +36,7 @@ void importTwiceWorksAsExpected() throws Exception {
         BibEntry expected = new BibEntry(StandardEntryType.InProceedings)
                 .withField(StandardField.AUTHOR, "1 ")
-                .withField(StandardField.TITLE, "Hello World")
+                .withField(StandardField.TITLE, "Hello World 1")
```
I check some paper in test data, all of title's characters should stay together as you said, so I add a `isFarAway` method to it to pass the unit test for hello world case.

koppor commented on src/test/resources/org/jabref/logic/importer/pdfContentImporterTest-pdfs.bib:
```diff
@@ -0,0 +1,110 @@
+@inproceedings{se2paper,
```
- The bib file should be in THE SAME directory as the .pdf files
- Set `.` as library-specific directory (in the library properties)
- Have a `file` attached to each entry in the `bib` file.
- Try to improve BibTeX data in the `bib` file:
  -  Add DOI
  - Fix authors
  - Add Year
  - Generate citation key
  - Rename PDF files to nicer name.

koppor reviewed (approved):
I need this functionality, therefore merging.

Please try to adress the comments in a follow-up PR.

koppor commented:
I added a CHANGELOG.md entry

koppor commented on src/test/java/org/jabref/logic/importer/fileformat/PdfContentImporterTest.java:
```diff
@@ -121,6 +126,29 @@ British Journal of Nutrition (2008), 99, 1–11 doi: 10.1017/S0007114507795296
+
+    @ParameterizedTest
+    @MethodSource("providePdfData")
+    void pdfTitleExtraction(String filePath, String expectedTitle) throws Exception {
```
Follow-up pull request: swap the arguments.

at `assert` in JUnit, first the expected value appears, then the input data. - This should also be done when wrapping inside `@ParameterizedTests`

koppor commented:
@leaf-soba I think, you need to do some cleanup of the filenames before a merge can go through. See test output on Windows test:

     Error: error: invalid path 'src/test/resources/pdfs/PdfContentImporter/peerj-cs-213 - On the impact of service-oriented patterns on software evolvability: a controlled experiment and metric-based analysis.pdf'

Sorry für that.

Maybe, you can adress all comments by me then.

leaf-soba commented:
Sorry I know it should be in another follow-up PR as you said before we should keep PR small but this one is too big, but I need to change the file name to pass the CI workflow, so it is a little hard to separate them.

> Set . as library-specific directory (in the library properties)
This one is only comment I didn't finish, since I don't understand it very clearly. Do you mean:
1. Open JabRef, load my BibTeX file.
2. Go to Library → Library Properties. 
3. In File Directory, enter a `.`
4. save my settings
But I don't know how to change it in code or project file as a PR, could you explain it in the follow-up PR.

koppor commented on src/test/resources/pdfs/PdfContentImporter/pdfContentImporterTest-pdfs.bib:
```diff
@@ -0,0 +1,128 @@
+@inproceedings{Kriha2018Teaching,
+author = {Walter Kriha and Tobias Jordine}
```
Note: This is INVALID BibTeX. All commans are missing.

In the future: Please try to open the file in JabRef.
